circuit-id-generator/tests/main.rs

```diff
@@ -0,0 +1,95 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use circuit_id_generator::circuit_id;
+use dusk_plonk::constraint_system::{ecc::*, StandardComposer};
+use dusk_plonk::jubjub::JubJubAffine;
+use dusk_plonk::prelude::*;
+
+// Implements a circuit that checks:
+// 1) a + b = c where C is a PI
+// 2) a <= 2^6
+// 3) b <= 2^5
+// 4) a * b = d where D is a PI
+// 5) JubJub::GENERATOR * e(JubJubScalar) = f where F is a PI
+#[derive(Debug, Default)]
+pub struct TestCircuit {
+    a: BlsScalar,
+    b: BlsScalar,
+    c: BlsScalar,
+    d: BlsScalar,
+    e: JubJubScalar,
+    f: JubJubAffine,
+}
+
+#[circuit_id]
+impl Circuit for TestCircuit {
+    fn gadget(&mut self, composer: &mut StandardComposer) -> Result<(), Error> {
+        let a = composer.add_input(self.a);
+        let b = composer.add_input(self.b);
+        let zero = composer.add_input(BlsScalar::zero());
+        // Make first constraint a + b = c
+        composer.poly_gate(
+            a,
+            b,
+            zero,
+            BlsScalar::zero(),
+            BlsScalar::one(),
+            BlsScalar::one(),
+            BlsScalar::zero(),
+            BlsScalar::zero(),
+            Some(-self.c),
+        );
+        // Check that a and b are in range
+        composer.range_gate(a, 1 << 6);
+        composer.range_gate(b, 1 << 5);
+        // Make second constraint a * b = d
+        composer.poly_gate(
+            a,
+            b,
+            zero,
+            BlsScalar::one(),
+            BlsScalar::zero(),
+            BlsScalar::zero(),
+            BlsScalar::one(),
+            BlsScalar::zero(),
+            Some(-self.d),
+        );
+
+        // This adds a PI also constraining `generator` to actually be
+        // `dusk_plonk::jubjub::GENERATOR`
+        let generator =
+            Point::from_public_affine(composer, dusk_plonk::jubjub::GENERATOR);
+        let e = composer.add_input(self.e.into());
+        let scalar_mul_result =
+            scalar_mul::variable_base::variable_base_scalar_mul(
+                composer, e, generator,
+            );
+        // Apply the constrain
+        composer.assert_equal_public_point(scalar_mul_result.into(), self.f);
+        Ok(())
+    }
+
+    fn padded_circuit_size(&self) -> usize {
+        1 << 11
+    }
+}
+
+mod tests {
+    use super::*;
+
+    #[test]
+    fn circuit_version_is_expected() {
+        assert_eq!(
+            &TestCircuit::CIRCUIT_ID,
+            &[
+                116, 100, 209, 255, 222, 124, 195, 235, 49, 246, 99, 176, 105,
+                168, 70, 123, 51, 188, 242, 98, 92, 8, 105, 144, 143, 139, 67,
+                25, 31, 177, 93, 188
+            ]
+        );
```

Comment:
The tests should checks:

1. Different structure with same circuit (same `gadget` and `padded_circuit_size`) ended up with the same Hash
2. Different structure with same `gadget` but different `padded_circuit_size` ended up with different hash
(we could also test different `gadget` but I think just checking one factor is enough)

If we use the generic approach:
1. Same structure different versioning ended up with different hash 
2. Test on code block that are not Circuit impl, with custom `const` name works